test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/ConnectionLimitTests.cs

```diff
@@ -0,0 +1,199 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure;
+using Microsoft.AspNetCore.Server.Kestrel.Tests;
+using Microsoft.AspNetCore.Testing;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.FunctionalTests
+{
+    public class ConnectionLimitTests
+    {
+        private static async Task KeepOpen(HttpContext context)
+        {
+            // begin the response
+            await context.Response.WriteAsync("Hello");
+
+            // keep it running until aborted
+            while (!context.RequestAborted.IsCancellationRequested)
```

Comment:
Instead of this why it just use a tcs that gets triggered on the cancellation callback?